torch/csrc/autograd/function.cpp

```diff
@@ -36,95 +36,59 @@ AnomalyMetadata* Function::metadata() noexcept {
   return anomaly_metadata_.get();
 }
 
-/*
- * Fix for #5534: prevent stack overflow on deletion of deep computation graph
- *
- * Sometimes one can end up with a very big computation graph of Functions
- * and Edges. Each std::shared_ptr<Function> contains a list of Edge, and
- * each Edge contains a std::shared_ptr<Function>. Deleting a
- * std::shared_ptr<Function> can trigger the recursive deletion of other
- * std::shared_ptr<Function>'s: this can stack overflow if the graph
- * is deep enough. Here is an example of such a graph:
- *
- * shared_ptr<Function> -> Edge -> shared_ptr<Function> -> Edge -> ... -> shared_ptr<Function>
- *
- * The solution here is to use a custom deleter with each
- * std::shared_ptr<Function>. The custom deleter keeps track of how many
- * nested deleters it is in. When this number exceeds the maximum allowed
- * depth, the Function* to be deleted are accumulated in a per-thread
- * delete queue and handled by one of the deleters.
- *
- * Note that these custom deleters are NOT necessary for deleting PyFunction.
- * This is because a THPFunction Python object owns a PyFunction that is in a
- * computation graph. When Python objects get recursively destroyed, they
- * are also queued into a delete list. This happens very early for them
- * (at 50 deleters): https://github.com/python/cpython/blob/f320be77ffb73e3b9e7fc98c37b8df3975d84b40/Include/object.h#L1024-L1063
- * so we don't need to worry about them.
- */
-
-thread_local std::deque<Function*> deleteFunctionQueue;
-thread_local size_t deleteFunctionRecursionDepth = 0;
-
-/*
- * If this number is set too high, a deep computation graph can still
- * stack overflow. The procedure for setting this number was to
- * 1) find the smallest value that would not guard against stack overflows
- *    on various machines
- * 2) Take the minimum of all such values and subtract some leeway because
- *    the memory of these stack frames will probably grow as time passes.
- * Testing on a few machines machines, the magic numbers were:
- * - Mac OSX (Macbook Pro 15) : ~60000
- * - A beefy Ubuntu 16.04 box : ~15000
- * - Windows AWS instance (g3.4xlarge): variable. My two attempts at different
- *   times have gotten the following numbers: ~8300, 3669
- */
-#ifdef _WIN32
-size_t deleteFunctionMaxRecursionDepth = 3000;
-#else
-size_t deleteFunctionMaxRecursionDepth = 10000;
-#endif
-
-struct RecursionDepthCounter {
- public:
-  explicit RecursionDepthCounter() {
-    ++deleteFunctionRecursionDepth;
-  }
-  ~RecursionDepthCounter() {
-    --deleteFunctionRecursionDepth;
+void gatherFunctions(Function* func,
+                     std::vector<std::shared_ptr<Function>>& stack) {
+  func->release_variables();
+  auto saved_vars = func->saved_variables();
```

Comment:
Actually can we remove this function altogether? It doesn't seem to be used anywhere, so this code will be dead.

Reply:
Function::saved_variables()?  I'm not familiar with its intended use, so don't feel qualified to say.  Maybe @goldsborough has an opinion on it?

Reply:
From what I can see, the code that used to use this function back when I re-factored large parts of the autograd in https://github.com/pytorch/pytorch/pull/5221 is now gone. In that case, this function is likely a legacy of how things used to be in the JIT and we could remove it

Reply:
Done.